### DISCUSSION TITLE
Rendering PC vector tiles with ipyleaflet/leafmap

### DISCUSSION THREAD

**Discussion by giswqs** [microsoft/PlanetaryComputer#175] [Q&A] [answered]
Excited to see the new release with vector tile support. I would like to add functionality to leafmap for supporting PC vector tiles. 
My question is what is the vector tile URL to a vector dataset? A URL like this does not work. 
`https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/United%20States_2022-06-14/tiles/{z}/{x}/{y}`

This example works (adapted from an ipyleaflet [example](https://ipyleaflet.readthedocs.io/en/latest/layers/vector_tile.html)):
```python
import leafmap
m = leafmap.Map()
url = 'https://tile.nextzen.org/tilezen/vector/v1/512/all/{z}/{x}/{y}.mvt?api_key=gCZXZglvRQa6sB2z7JzL1w'
attribution = "Nextzen"
vector_tile_layer_styles = {}
m.add_vector_tile_layer(url, attribution, vector_tile_layer_styles)
m
```
![image](https://user-images.githubusercontent.com/5016453/215007863-3b3e5de7-9eec-4466-9fac-fc5862ab89f7.png)

This example does not work. The vector tile does not render
```python
m = leafmap.Map()
url = 'https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/United%20States_2022-06-14/tiles/{z}/{x}/{y}'
attribution = "Microsoft"
vector_tile_layer_styles = {}
m.add_vector_tile_layer(url, attribution, vector_tile_layer_styles)
m
```
![image](https://user-images.githubusercontent.com/5016453/215008340-a8115fe9-d903-499d-a3ae-88ea74d54d36.png)

**Comment by TomAugspurger** [comment, marked as answer]
Thanks @giswqs. I think that this should work for you:

```
import leafmap
m = leafmap.Map()
url = "https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/{z}/{x}/{y}"
attribution = "Microsoft"
vector_tile_layer_styles = {}
m.add_vector_tile_layer(url, attribution, vector_tile_layer_styles)
m
```

For this specific dataset, the tiles are a collection-level asset. You can get that URL programatically with

```python
import requests
import pystac_client

catalog = pystac_client.Client.open("https://planetarycomputer.microsoft.com/api/stac/v1/")

asset_href = catalog.get_collection("ms-buildings").assets["global-footprints"].href
url = requests.get(asset_href).json()["tiles"][0]
```

When zoomed all the way out things seem a bit slow to update, but it gets better when zoomed in.

**Comment by giswqs** [comment]
@TomAugspurger Thanks for providing the URL. It works now. One issue I notice is that the tile does not render when zoom level >13. I guess this is an ipyleaflet issue. Adding parameters like `max_native_zoom`, `maxNativeZoom`, `max_zoom`, `maxZoom` has no effect. I need to look into it a bit more.  

```python
import ipyleaflet
m = ipyleaflet.Map(center=(40, -100), zoom=4, scroll_wheel_zoom=True)
url = 'https://planetarycomputer.microsoft.com/api/data/v1/vector/collections/ms-buildings/tilesets/global-footprints/tiles/{z}/{x}/{y}'
layer = ipyleaflet.VectorTileLayer(url=url, attribution='Microsoft', max_native_zoom=20, max_zoom=20)
m.add(layer)
m
```
![Peek 2023-01-27 09-13](https://user-images.githubusercontent.com/5016453/215107809-3783a949-7d21-4090-bc5c-634512b1c0ad.gif)
